linera-execution/src/wasm/async_boundary.rs

```diff
@@ -5,156 +5,240 @@
 //! modules.
 
 use super::{
+    async_determinism::HostFutureQueue,
     common::{ApplicationRuntimeContext, WasmRuntimeContext},
-    ExecutionError,
+    ExecutionError, WasmExecutionError,
 };
-use futures::{future::BoxFuture, ready, stream::StreamExt};
+use futures::{channel::oneshot, ready, stream::StreamExt, FutureExt};
 use std::{
-    any::type_name,
-    fmt::{self, Debug, Formatter},
     future::Future,
     marker::PhantomData,
     pin::Pin,
-    sync::Arc,
-    task::{Context, Poll, Waker},
+    task::{Context, Poll},
+    thread,
 };
-use tokio::sync::Mutex;
 
-/// A host future that can be called by a Wasm guest module.
-pub struct HostFuture<'future, Output> {
-    future: Mutex<BoxFuture<'future, Output>>,
+/// An actor that runs a future implemented in a Wasm module.
+///
+/// The actor should run in its own non-asynchronous thread, where it will block until it receives
+/// poll requests from a [`PollSender`].
+pub struct GuestFutureActor<Future, Application>
+where
+    Application: ApplicationRuntimeContext,
+    Future: GuestFutureInterface<Application>,
+{
+    context: WasmRuntimeContext<Application>,
+    poll_requests: std::sync::mpsc::Receiver<oneshot::Sender<PollResponse<Future::Output>>>,
```

Comment:
I'd like to keep the full path, because there are asynchronous and synchronous `mpsc` channels, and here the synchronous channels is mixed inside an "asynchronous type", so I wanted to make that explicit as an extra precaution.